miniwin/miniwin/src/miniwin_d3drmmesh.cpp

```diff
@@ -38,85 +62,158 @@ HRESULT Direct3DRMMeshImpl::GetGroup(
+		return DDERR_INVALIDPARAMS;
+	}
+
+	if (vertexCount) {
+		*vertexCount = static_cast<DWORD>(it->second.vertices.size());
+	}
+	if (faceCount) {
+		*faceCount = static_cast<DWORD>(it->second.faces.size() / it->second.vertexPerFace);
+	}
+	if (vertexPerFace) {
+		*vertexPerFace = static_cast<DWORD>(it->second.vertexPerFace);
+	}
+	if (dataSize) {
+		*dataSize = static_cast<DWORD>(it->second.faces.size());
+	}
+	if (data && it != m_groups.end()) {
+		unsigned int* outData = reinterpret_cast<unsigned int*>(data);
+		std::copy(it->second.faces.begin(), it->second.faces.end(), outData);
+	}
+
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::SetGroupColor(int groupIndex, D3DCOLOR color)
 {
-	MINIWIN_NOT_IMPLEMENTED();
+	m_groups[groupIndex].color = color;
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::SetGroupColorRGB(int groupIndex, float r, float g, float b)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	return DD_OK;
-}
+	r = std::clamp(r, 0.0f, 1.0f);
+	g = std::clamp(g, 0.0f, 1.0f);
+	b = std::clamp(b, 0.0f, 1.0f);
 
-HRESULT Direct3DRMMeshImpl::SetGroupMaterial(int groupIndex, IDirect3DRMMaterial* material)
-{
-	MINIWIN_NOT_IMPLEMENTED();
-	return DD_OK;
-}
+	D3DCOLOR color = (0xFF << 24) | (static_cast<BYTE>(r * 255.0f) << 16) | (static_cast<BYTE>(g * 255.0f) << 8) |
+					 (static_cast<BYTE>(b * 255.0f));
 
-HRESULT Direct3DRMMeshImpl::SetGroupMapping(D3DRMGROUPINDEX groupIndex, D3DRMMAPPING mapping)
-{
-	MINIWIN_NOT_IMPLEMENTED();
+	m_groups[groupIndex].color = color;
 	return DD_OK;
 }
 
-HRESULT Direct3DRMMeshImpl::SetGroupQuality(int groupIndex, D3DRMRENDERQUALITY quality)
+D3DCOLOR Direct3DRMMeshImpl::GetGroupColor(D3DRMGROUPINDEX index)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	return DD_OK;
+	auto it = m_groups.find(index);
+	return (it != m_groups.end()) ? it->second.color : 0xFFFFFFFF;
 }
 
-HRESULT Direct3DRMMeshImpl::SetVertices(int groupIndex, int offset, int count, D3DRMVERTEX* vertices)
+HRESULT Direct3DRMMeshImpl::SetGroupMaterial(int groupIndex, IDirect3DRMMaterial* material)
 {
-	MINIWIN_NOT_IMPLEMENTED();
+	m_groups[groupIndex].material = material;
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::SetGroupTexture(int groupIndex, IDirect3DRMTexture* texture)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	m_groupTexture = texture;
+	m_groups[groupIndex].texture = texture;
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::GetGroupTexture(int groupIndex, LPDIRECT3DRMTEXTURE* texture)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	if (!m_groupTexture) {
+	auto it = m_groups.find(groupIndex);
+	if (it == m_groups.end() || !it->second.texture) {
 		return DDERR_GENERIC;
 	}
 
-	m_groupTexture->AddRef();
-	*texture = m_groupTexture;
+	it->second.texture->AddRef();
+	*texture = it->second.texture;
+	return DD_OK;
+}
+
+HRESULT Direct3DRMMeshImpl::SetGroupMapping(D3DRMGROUPINDEX groupIndex, D3DRMMAPPING mapping)
+{
 	return DD_OK;
 }
 
 D3DRMMAPPING Direct3DRMMeshImpl::GetGroupMapping(int groupIndex)
 {
-	MINIWIN_NOT_IMPLEMENTED();
 	return D3DRMMAP_PERSPCORRECT;
```

Comment:
Let's also make this a member of `MeshGroup`, and initialize it to `D3DRMMAP_PERSPCORRECT`?

Reply:
Why D3DRMMAP_PERSPCORRECT is the default and the only option :)

Reply:
OK then :)